### DISCUSSION TITLE
What are additional LetPattern and MatchPattern cases that'd be useful?

### DISCUSSION THREAD

**Discussion by StachuDotNet** [darklang/dark#4714] [Suggestions]
Useful links:
- https://fsharp.github.io/fsharp-compiler-docs/reference/fsharp-compiler-syntax-synpat.html
- https://learn.microsoft.com/en-us/dotnet/fsharp/language-reference/pattern-matching#remarks

**Comment by StachuDotNet** [comment]
should this ever be legal Dark code?
```fsharp
match [1; 2; 3] with
| first :: second :: [] -> "exactly 2 elements"
| _ -> "not 2 elements"
```

**Reply by pbiggar** [reply]
I'd probably think that something based off elipses is better (like rescript does for records). Something like:

```
match [1,2,3]
| [first, second, ...] -> two or more
| [first, second] -> exactly two
| etc
```

**Comment by StachuDotNet** [comment]
should this ever be legal Dark code? (relevant to #4711 right now) 
`let (a, b, (c, d)) = (1, 2, (3, 4))`

**Reply by StachuDotNet** [reply]
For what it's worth, I think we could support it in the backend, and limit the 'depth' based on something other than the model of the language. (and in the meantime, possibly restrict it altogether in the editor)
For example, an individual canvas could disallow a pattern like this based on stylistic preferences.

**Reply by pbiggar** [reply]
I think this is fine, not even sure why we'd restrict it (sometimes you have data in that shape!)

**Comment by StachuDotNet** [comment]
Should this be legal? ("constructor"/enum let pattern)
```fsharp
let opt = Just 1
let (Just a) = opt
a
```

**Reply by pbiggar** [reply]
No.

I think in general the following shoudl be legal, but there's issues. Suppose we have the type: `type clientID = ClientID of int`

Then it makes sense to have: `let (ClientID id) = x`

But then if you expand the type to `type clientID = ClientID of int | NewClientID of int64`

Then deconstructing a single variant/constructor becomes a type error suddenly, which is bad.

So I'd propose we don't allow deconstruction of constructors at all right now.
